### PR TITLE
listener: fix flaky listener integration test

### LINKED ISSUE #22348: `//test/integration:listener_lds_integration_test` is flaky or failing in CI
i think this is a relatively recent issue - failing CI can be seen in the check/asan job as here:

https://dev.azure.com/cncf/envoy/_build/results?buildId=113433&view=logs&j=1439b9f7-a348-5b50-b5fe-ea612ea91241&t=37b0a9be-1a71-50b3-594a-4c04a031247d&l=900

i think the relevant part of log is:

```console

Value of: codec->connected()
  Actual: true
Expected: false
Stack trace:
... Google Test internal frames ...

test/integration/listener_lds_integration_test.cc:827: Failure
Value of: codec->connection()->transportFailureReason()
Expected: starts with "delayed connect error"
  Actual: "" (of type absl::string_view)
Stack trace:
  0x45a5639: Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveListenerAfterInPlaceUpdate_Test::TestBody()
  0x45a72c0: Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveListenerAfterInPlaceUpdate_Test::TestBody()
  0xd85d0e5: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0xd825003: testing::Test::Run()
  0xd826b68: testing::TestInfo::Run()
... Google Test internal frames ...

[2022-07-21 10:14:32.249][16][critical][assert] [source/common/network/connection_impl.cc:107] assert failure: !ioHandle().isOpen() && delayed_close_timer_ == nullptr. Details: ConnectionImpl was unexpectedly torn down without being closed.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==16==ERROR: AddressSanitizer: ABRT on unknown address 0xfffe00000010 (pc 0x7facf9406e87 bp 0x7ffe006360b0 sp 0x7ffe00635c70 T0)
    #0 0x7facf9406e87 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x3ee87) (BuildId: 71f0f3074a929e519e85f6a5c03a7d1fd976bfe4)
    #1 0x7facf94087f0 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x407f0) (BuildId: 71f0f3074a929e519e85f6a5c03a7d1fd976bfe4)
    #2 0xa93a111 in Envoy::Network::ConnectionImpl::~ConnectionImpl() /proc/self/cwd/source/common/network/connection_impl.cc:106:3
    #3 0xa97f44d in virtual thunk to Envoy::Network::ClientConnectionImpl::~ClientConnectionImpl() /proc/self/cwd/./source/common/network/connection_impl.h:259:7
    #4 0x470eb89 in std::__1::unique_ptr<Envoy::Network::ClientConnection, std::__1::default_delete<Envoy::Network::ClientConnection> >::reset(Envoy::Network::ClientConnection*) /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:54:5
    #5 0x47b1e14 in Envoy::Http::CodecClient::~CodecClient() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:269:19
    #6 0x47bcd0d in Envoy::IntegrationCodecClient::~IntegrationCodecClient() /proc/self/cwd/./test/integration/http_integration.h:28:7
    #7 0x46b4ff9 in std::__1::unique_ptr<Envoy::IntegrationCodecClient, std::__1::default_delete<Envoy::IntegrationCodecClient> >::reset(Envoy::IntegrationCodecClient*) /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:54:5
    #8 0x45a5804 in Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveListenerAfterInPlaceUpdate_Test::TestBody() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:269:19
    #9 0x45a72bf in non-virtual thunk to Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveListenerAfterInPlaceUpdate_Test::TestBody() /proc/self/cwd/test/integration/listener_lds_integration_test.cc
    #10 0xd85d0e4 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #11 0xd825002 in testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2655:5
    #12 0xd826b67 in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2832:11
    #13 0xd828f04 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #14 0xd84b2ba in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #15 0xd8610e3 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #16 0xd849fc6 in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #17 0xa668834 in Envoy::TestRunner::RunTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #18 0xa663f87 in main /proc/self/cwd/test/main.cc:34:10
    #19 0x7facf93e9c86 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21c86) (BuildId: 71f0f3074a929e519e85f6a5c03a7d1fd976bfe4)
    #20 0x449a9a9 in _start (/b/f/w/bazel-out/k8-dbg/bin/test/integration/listener_lds_integration_test.runfiles/envoy/test/integration/listener_lds_integration_test+0x449a9a9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: ABRT (/lib/x86_64-linux-gnu/libc.so.6+0x3ee87) (BuildId: 71f0f3074a929e519e85f6a5c03a7d1fd976bfe4) in raise
==16==ABORTING


```

according to azp analytics this test has failed 2/2 times in the last 14 days (on postsubmit, its also happening on presubmit):

https://dev.azure.com/cncf/envoy/_test/analytics?definitionId=11&contextType=build

### PR DESCRIPTION
**Pull request by soulxu**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>

Commit Message: listener: fix flaky listener integration test
Additional Description:
The listener socket is closed async https://github.com/envoyproxy/envoy/blob/3f0efa989f9776a8e1fcff3b1d1464da66aecb7e/source/server/listener_manager_impl.cc#L561-L567

So there can be a chance the listener was removed, but the socket doesn't remove yet.

Risk Level: low
Testing: integration test
Docs Changes: n/a 
Release Notes: n/a
Platform Specific Features: n/a
Fixes #22348

### CODE REVIEW

File: test/integration/listener_lds_integration_test.cc [hunk 1]
```diff
@@ -821,10 +821,18 @@ TEST_P(ListenerIntegrationTest, RemoveListenerAfterInPlaceUpdate) {
...
   // All the listen socket are closed. include the sockets in the active listener and
   // the sockets in the filter chain draining listener. The new connection should be reset.
-  auto codec =
+  auto codec1 =
       makeRawHttpConnection(makeClientConnection(lookupPort(listener_name_)), absl::nullopt);
-  EXPECT_FALSE(codec->connected());
-  EXPECT_THAT(codec->connection()->transportFailureReason(), StartsWith("delayed connect error"));
+  // The socket are closed asynchronously, so waiting the connection closed here.
+  while (codec1->connected()) {
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
```
Comment (jmarantz) [inline test/integration/listener_lds_integration_test.cc:828]: can this be done with a waitFor variant? It looks like this pattern will spin. Maybe waitForDisconnect in integration_tcp_client.cc ?

there's one below as well.

Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:828]: yea, waitForDisconnect is right thing, thanks!

**Comment by jmarantz** [comment]
Yan -- can you do a senior maintainer pass?

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**